cmd/nvidia-dra-controller/imex.go

```diff
@@ -102,35 +114,44 @@ func StartIMEXManager(ctx context.Context, config *Config) (*ImexManager, error)
 }
 
 // manageResourceSlices reacts to added and removed IMEX domains and triggers the creation / removal of resource slices accordingly.
-func (m *ImexManager) manageResourceSlices(ctx context.Context, owner resourceslice.Owner, addedDomainsCh <-chan string, removedDomainsCh <-chan string) error {
-	driverResources := &resourceslice.DriverResources{}
-	controller, err := resourceslice.StartController(ctx, m.clientset, DriverName, owner, driverResources)
+func (m *ImexManager) manageResourceSlices(ctx context.Context) error {
+	klog.Info("Start streaming IMEX domains from nodes...")
+	addedDomainsCh, removedDomainsCh, err := m.streamImexDomains(ctx)
+	if err != nil {
+		return fmt.Errorf("error streaming IMEX domains: %w", err)
+	}
+
+	klog.Info("Start publishing IMEX channels to ResourceSlices...")
+	controller, err := resourceslice.StartController(ctx, m.clientset, m.driverName, m.owner, m.driverResources)
 	if err != nil {
 		return fmt.Errorf("error starting resource slice controller: %w", err)
 	}
 
-	imexDomainOffsets := new(imexDomainOffsets)
 	m.waitGroup.Add(1)
 	go func() {
 		defer m.waitGroup.Done()
 		for {
 			select {
 			case addedDomain := <-addedDomainsCh:
-				offset, err := imexDomainOffsets.add(addedDomain, ResourceSliceImexChannelLimit, DriverImexChannelLimit)
-				if err != nil {
-					klog.Errorf("Error calculating channel offset for IMEX domain %s: %v", addedDomain, err)
-					return
-				}
 				klog.Infof("Adding channels for new IMEX domain: %v", addedDomain)
-				driverResources := driverResources.DeepCopy()
-				driverResources.Pools[addedDomain] = generateImexChannelPool(addedDomain, offset, ResourceSliceImexChannelLimit)
-				controller.Update(driverResources)
+				if err := m.addImexDomain(addedDomain); err != nil {
+					klog.Errorf("Error adding channels for IMEX domain %s: %v", addedDomain, err)
+					go func() {
+						time.Sleep(m.retryTimeout)
```

Comment:
What if the error is a malformed label value? we would retry indefinitely. should we have error types?

Reply:
Hmm, true. Let me think about it. We should also technically be canceling the sleep if we ever get a ctx.Done(), but I reasoned that it wasn't worth the complexity. If we need to add logic to handle an error type, then maybe its worth a helper function to do the sleep/repush into the channel, at which point I should also make it robust to exiting early on ctx.Done().

Reply:
Pushed an update